### ISSUE TITLE
Failure to handle 130MB LZMA2-24 7zAES encrypted archive created on Ubuntu Linux

### ISSUE THREAD

**Issue by pulsar70** [miurahr/py7zr#272]
with py7zr.SevenZipFile(pathname, 'r', password=pwd) as archive:

throws py7zr.exceptions.Bad7zFile: invalid block data

--> the archive is OK and the provided password is correct (checked on 7-zip executable on Windows 10).
--> password is 40 chars long, ASCII only.

Any idea where the problem is?

**Comment by miurahr** [comment]
Could you please post all the stack trace and log for the incident?

**Comment by pulsar70** [comment]
In Windows Powershell : no apparent verbose output :

PS py7zr x --verbose -P .\premium_20201104_partition_00005000_00000000.7z
Password:
Header is corrupted. Cannot read as 7z file.


Using the Python API :

Processing file [premium_20201104_partition_00005000_00000000.7z]
Traceback (most recent call last):
  File "XXX\extract_demographics.py", line 30, in <module>
    with py7zr.SevenZipFile(pathname, 'r', password=pwd) as archive:
  File "C:\Python364_64\lib\site-packages\py7zr\py7zr.py", line 327, in __init__
    raise e
  File "C:\Python364_64\lib\site-packages\py7zr\py7zr.py", line 313, in __init__
    self._real_get_contents(password)
  File "C:\Python364_64\lib\site-packages\py7zr\py7zr.py", line 356, in _real_get_contents
    header = Header.retrieve(self.fp, buffer, self.afterheader, password)
  File "C:\Python364_64\lib\site-packages\py7zr\archiveinfo.py", line 862, in retrieve
    obj._read(fp, buffer, start_pos, password)
  File "C:\Python364_64\lib\site-packages\py7zr\archiveinfo.py", line 905, in _read
    raise Bad7zFile('invalid block data')
py7zr.exceptions.Bad7zFile: invalid block data

**Comment by miurahr** [comment]
The error happened as a reason for CRC error.

Here is around line 905 in archiveinfo.py

```
            if folder.digestdefined:
                if folder.crc != calculate_crc32(folder_data):
                    raise Bad7zFile('invalid block data')
```

This means all the data py7zr recognized have been extracted.  
If data is correct, CRC logic are broken.
If data is broken, password may be wrong, or the key calculated from password may be wrong, or noother is wrong.

**Comment by miurahr** [comment]
Is the password consist of only an ascii characters?
I don't test it with other than ascii characters.

**Comment by pulsar70** [comment]
Again : 
1) The same archive with the same password can be process without problem by 7-zip executable (www.7zip.org). Therefore data is not broken.
2) Password is pure ASCII characters (length : 40). Only uppercase, lowercase english alphabet + digits, no special characters.

**Comment by miurahr** [comment]
For clarification.
Is the extracted data broken? Is it completely broken or difference at some part of data?
If it is the former, it would be a issue on a  generation of secret key for decryption, the latter indicate that a decompression algorithm may have a problem.

**Comment by miurahr** [comment]
And could you provide  information issue template asked?

---
To Reproduce Steps to reproduce the behavior: (example)

Prepare test data attached as 'file' in current directory.
Run following code with python3. ' import py7zr archive = py7zr.SevenZipFile('test.7z','w') archive.writeall('file') archive.close() '
test.7z is produced
run command in ternminal '7z l test.7z'
See error
Expected behavior A clear and concise description of what you expected to happen.

Environment (please complete the following information):

OS: [e.g. Windows 10, Ubuntu Linux 18.04.01]
Python [e.g. 3.6, pypy3.6.9-7.3.0]
py7zr version: [e.g. v0.5b1, commit #123456 on master]
Test data(please attach in the report): A minimum test data to reproduce your problem.

Additional context Add any other context about the problem here.

**Comment by pulsar70** [comment]
No extracted data is present, even when I specify a target output directory.

py7zr Version 0.11.0 : Copyright (C) 2019,2020 Hiroshi Miura (Python 3.6.4 [CPython MSC v.1900 32 bit (Intel)])

Windows 10 Enterprise

Cannot provide test data, the problem arises on my 130MB production archive containing business confidential data...

**Comment by miurahr** [comment]
> File "C:\Python364_64\lib\site-packages\py7zr\py7zr.py", line 356, in _real_get_contents
header = Header.retrieve(self.fp, buffer, self.afterheader, password)

From stack trace, a problem seems  happen at header decryption.
And you seems use python 3.6.4(64bit).

**Comment by pulsar70** [comment]
The stack trace came from my usage of py7zr API with Python 64 bits.

The line 
py7zr Version 0.11.0 : Copyright (C) 2019,2020 Hiroshi Miura (Python 3.6.4 [CPython MSC v.1900 32 bit (Intel)])
comes from using your command line tool in Powershell.

Both uses fails to decrypt and extract my archive.

**Comment by miurahr** [comment]
What algorithm set does the 7z file **header**  encrypted with?
We have several test cases for encrypted files.
https://github.com/miurahr/py7zr/blob/master/tests/test_encryption.py

for example, `encrypted_3.7z` is 
```
$ 7z l encrypted_3.7z 

7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=ja_JP.UTF-8,Utf16=on,HugeFiles=on,64 bits,16 CPUs AMD Ryzen 7 PRO 4750U with Radeon Graphics (860F01),ASM,AES-NI)

Scanning the drive for archives:
1 file, 711 bytes (1 KiB)

Listing archive: encrypted_3.7z


Enter password (will not be echoed):
--
Path = encrypted_3.7z
Type = 7z
Physical Size = 711
Headers Size = 263
Method = LZMA2:12 7zAES
Solid = +
Blocks = 1

   Date      Time    Attr         Size   Compressed  Name
------------------- ----- ------------ ------------  ------------------------
2019-03-14 09:10:08 D....            0            0  scripts
2019-03-14 09:10:08 ....A          111          448  scripts/py7zr
2019-03-14 09:07:13 ....A           58               setup.cfg
2019-03-14 09:09:01 ....A          559               setup.py
------------------- ----- ------------ ------------  ------------------------
2019-03-14 09:10:08                728          448  3 files, 1 folders

```

Encrypted with 7zAES + LZMA2 with header encryption. This test case has passed.

**Comment by pulsar70** [comment]
7-Zip [64] 16.04 : Copyright (c) 1999-2016 Igor Pavlov : 2016-10-04

Scanning the drive for archives:
1 file, 136897204 bytes (131 MiB)

Listing archive: d:\Facepoint\FPMaster_dump\premium_20201104_partition_00005000_00000000.7z

--
Path = d:\Facepoint\FPMaster_dump\premium_20201104_partition_00005000_00000000.7z
Type = 7z
Physical Size = 136897204
Headers Size = 312212
Method = LZMA2:24 7zAES
Solid = +
Blocks = 1

**Comment by miurahr** [comment]
There is no strange things. If there is any special, a header size is large.

> Headers Size = 312212

A I/O block  size in  python 3.6 is  32kB.
If you  tested it with pytohn 3.7.5 and later, it would be 1MB.

This is because LZMA library in python 3.6, and 3.7.4 and before has a known bug.

py7zr/properties.py:
```
if platform.python_implementation() == "PyPy" and sys.version_info >= (3, 6, 9):
    READ_BLOCKSIZE = 1048576  # type: int
elif sys.version_info >= (3, 7, 5):
    READ_BLOCKSIZE = 1048576
else:
    READ_BLOCKSIZE = 32768
```

**Comment by miurahr** [comment]
Python 3.6.4 may have a bug in lzma module.
https://bugs.python.org/issue21872

A fix has been merged in 3.7, 3.8 and 3.9.

**Comment by pulsar70** [comment]
Thank you Hiroshi, I'll try to upgrade to 3.7+ and will let you know if it changes anything.

Jerome
